### PR TITLE
Checking Validity of all Certs in the chain [VAULT-2114]

### PR DESCRIPTION
**Pull request by hghaf099**

### CODE REVIEW

File: vault/diagnose/tls_verification.go [hunk 1]
```diff
@@ -136,11 +176,25 @@ func ParseTLSInformation(certFilePath string) ([]*x509.Certificate, []*x509.Cert
 			leafCerts = append(leafCerts, cert)
 		}
 	}
+
+	rootSubjects := GetCertsSubjects(rootCerts)
```
Comment (hghaf099) [inline vault/diagnose/tls_verification.go:180]: I am not sure if this is needed or not! For self-signed certs, the cert can verify itself, and the subject and issuer are the same. So, just checking if we have subjects is not enough to determine if the cert is a self-signed cert. So, maybe we can just check if  "len(rootCerts) == 0".

File: vault/diagnose/tls_verification.go [hunk 2]
```diff
@@ -185,6 +248,11 @@ func TLSErrorChecks(leafCerts, interCerts, rootCerts []*x509.Certificate) error
 // and root certificates provided.
 func TLSFileWarningChecks(leafCerts, interCerts, rootCerts []*x509.Certificate) ([]string, error) {
 	var warnings []string
+	// add a warning for when there are more than one leaf certs
+	if len(leafCerts) > 1 {
+		warnings = append(warnings, fmt.Sprintf("LeafCerts contains more than one leafCert."))
```
Comment (HridoyRoy) [inline vault/diagnose/tls_verification.go:253]:
```suggestion
		warnings = append(warnings, fmt.Sprintf("More than one leaf certificate detected. Please ensure that there is one unique leaf certificate being supplied to vault in the vault server config file."))
```

File: vault/diagnose/tls_verification_test.go [hunk 3]
```diff
@@ -117,7 +117,7 @@ func TestTLSExpiredCert(t *testing.T) {
 	if !strings.Contains(errs[0].Error(), "certificate has expired or is not yet valid") {
 		t.Fatalf("Bad error message: %s", errs[0])
 	}
-	if warnings == nil || len(warnings) != 1 {
```
Comment (HridoyRoy) [inline vault/diagnose/tls_verification_test.go:120]: Can you add a check for the other warning thrown in this test to ensure it is expected?

File: vault/diagnose/tls_verification_test.go [hunk 4]
```diff
@@ -201,7 +201,7 @@ func TestTLSMultiKeys(t *testing.T) {
 }
 
 // TestTLSMultiCerts verifies that a unique error message is thrown when a cert is specified twice.
```
Comment (HridoyRoy) [inline vault/diagnose/tls_verification_test.go:203]:
```suggestion
// TestTLSCertAsKey verifies that a unique error message is thrown when a cert is specified twice.
```

File: vault/diagnose/tls_verification_test.go [hunk 5]
```diff
@@ -276,6 +276,27 @@ func TestTLSNoRoot(t *testing.T) {
...
+				TLSCertFile:                   "./test-fixtures/selfSignedCert.pem",
+				TLSKeyFile:                    "./test-fixtures/selfSignedCertKey.pem",
+				TLSMinVersion:                 "tls10",
+				TLSRequireAndVerifyClientCert: true,
+				TLSDisableClientCerts:         false,
+			},
+		},
+	}
+	_, errs := ListenerChecks(context.Background(), listeners)
+	if errs != nil {
```
Comment (HridoyRoy) [inline vault/diagnose/tls_verification_test.go:295]: Could you do an explicit check here to make sure the right error is returned for this test?

**Review by HridoyRoy — CHANGES_REQUESTED** [review]
Looking very good so far! Just a few small changes and then it's good to go!

**Comment by HridoyRoy** [comment]
Seems like `test-go` checks are failing due to the new code. Could you update those tests to reflect the code changes you've made?

File: vault/diagnose/tls_verification_test.go [hunk 6]
```diff
@@ -146,9 +147,13 @@ func TestTLSMismatchedCryptographicInfo(t *testing.T) {
 	if errs == nil || len(errs) != 1 {
 		t.Fatalf("TLS Config check on fake certificate should fail")
 	}
-	if !strings.Contains(errs[0].Error(), "tls: private key type does not match public key type") {
+	if !strings.Contains(errs[0].Error(), "certificate signed by unknown authority") {
 		t.Fatalf("Bad error message: %s", errs[0])
 	}
+	_, err := tls.LoadX509KeyPair("./../../api/test-fixtures/keys/cert.pem", "./test-fixtures/ecdsa.key")
```
Comment (hghaf099) [inline vault/diagnose/tls_verification_test.go:153]: The ListenerChecks returns with error before it tls.LoadX509KeyPair in TLSFileChecks

File: vault/diagnose/tls_verification_test.go [hunk 7]
```diff
@@ -276,6 +287,30 @@ func TestTLSNoRoot(t *testing.T) {
 	}
 }
 
+func TestTLSSelfSignedCert(t *testing.T) {
```
Comment (hghaf099) [inline vault/diagnose/tls_verification_test.go:290]: I should have removed this test, I will add that to the other PR

File: vault/diagnose/tls_verification.go [hunk 8]
```diff
@@ -52,23 +52,63 @@ func ListenerChecks(ctx context.Context, listeners []listenerutil.Listener) ([]s
...
+		}
+	}
+	return false
+}
+
+// GetCertsSubjects returns a list of the DER-encoded subjects of
+// all of the certificates in the slice.
+func GetCertsSubjects(certs []*x509.Certificate) [][]byte {
+	res := make([][]byte, len(certs))
+	for i, lc := range certs {
```
[the remark this answers is not in the document]
Comment (hghaf099) [reply vault/diagnose/tls_verification.go:94]: did not get what needs to be done here? should I change the naming for "lc"? :)

[the remark this answers is not in the document]
Comment (HridoyRoy) [reply vault/diagnose/tls_verification.go:94]: Sorry about the confusion! I meant to say, let's change the name `lc` to `leafCert`

**Review by HridoyRoy — APPROVED** [review]
Looks good, nice work!

**Review by HridoyRoy — APPROVED** [review]

**Pull request merged**